core/src/test/java/io/questdb/griffin/SnapshotTest.java

```diff
@@ -409,6 +553,135 @@ public void testSnapshotUnknownSubOptionFails() throws Exception {
+                    " from long_sequence(5)" +
+                    ") timestamp(ts) partition by DAY WAL");
+
+            assertWalExistence(true, tableName, 1);
+            assertSegmentExistence(true, tableName, 1, 0);
+
+            drainWalQueue();
+
+            assertWalExistence(true, tableName, 1);
+
+            assertSql(tableName, "x\tts\n" +
+                    "1\t2022-02-24T00:00:00.000000Z\n" +
+                    "2\t2022-02-24T00:00:01.000000Z\n" +
+                    "3\t2022-02-24T00:00:02.000000Z\n" +
+                    "4\t2022-02-24T00:00:03.000000Z\n" +
+                    "5\t2022-02-24T00:00:04.000000Z\n");
+
+            WalPurgeJobTest.MicrosecondClockMock clock = new WalPurgeJobTest.MicrosecondClockMock();
+            final long interval = engine.getConfiguration().getWalPurgeInterval() * 1000;
+            FilesFacade ff = configuration.getFilesFacade();
+            WalPurgeJob job = new WalPurgeJob(engine, ff, clock);
+            snapshotAgent.setWalPurgeJobRunLock(job.getRunLock());
+
+            compiler.compile("snapshot prepare", sqlExecutionContext);
+            Thread controlThread1 = new Thread(() -> {
+                clock.timestamp = interval;
+                while (job.run(0)) {
+                    // run until empty
+                }
+                Path.clearThreadLocals();
+            });
+
+            controlThread1.start();
+            controlThread1.join();
+
+            assertSegmentExistence(true, tableName, 1, 0);
+            assertWalExistence(true, tableName, 1);
+
+            engine.releaseInactive();
+
+            compiler.compile("snapshot complete", sqlExecutionContext);
+            Thread controlThread2 = new Thread(() -> {
+                clock.timestamp = 2 * interval;
+                while (job.run(0)) {
+                    // run until empty
+                }
+                Path.clearThreadLocals();
+            });
+
+            controlThread2.start();
+            controlThread2.join();
+
+            job.close();
+            snapshotAgent.setWalPurgeJobRunLock(null);
+
+            assertSegmentExistence(false, tableName, 1, 0);
+            assertWalExistence(false, tableName, 1);
+        });
+    }
+
+    @Test
+    public void testWalMetadataRecovery() throws Exception {
+        final String snapshotId = "id1";
+        final String restartedId = "id2";
+        assertMemoryLeak(() -> {
+            snapshotInstanceId = snapshotId;
+            String tableName = testName.getMethodName() + "_abc";
+            compile("create table " + tableName + " as (" +
+                    "select x, " +
+                    " rnd_symbol('AB', 'BC', 'CD') sym, " +
+                    " timestamp_sequence('2022-02-24', 1000000L) ts, " +
+                    " rnd_symbol('DE', null, 'EF', 'FG') sym2 " +
+                    " from long_sequence(5)" +
+                    ") timestamp(ts) partition by DAY WAL");
+
```

Comment:
Please add test case with few alters before snapshot taken so that structure version is not 0.
Also add a noop `update` which does not change any rows.